File: src/main/java/com/music/music_inventory_api/service/CustomerService.java
package com.music.music_inventory_api.service;

import com.music.music_inventory_api.dto.request.CreateCustomerRequest;
import com.music.music_inventory_api.dto.request.UpdateCustomerRequest;
import com.music.music_inventory_api.dto.response.CustomerResponse;
import com.music.music_inventory_api.dto.response.OrderResponse;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/** Service interface for Customer operations. */
public interface CustomerService
{
    /** Create a new customer. */
    CustomerResponse createCustomer(CreateCustomerRequest request);

    /** Get customer by ID. */
    CustomerResponse getCustomerById(Long id);

    /** Get customer by email. */
    CustomerResponse getCustomerByEmail(String email);

    /** Get all customers with pagination. */
    Page<CustomerResponse> getAllCustomers(Pageable pageable);

    /** Update an existing customer. */
    CustomerResponse updateCustomer(Long id, UpdateCustomerRequest request);

    /** Delete a customer. */
    void deleteCustomer(Long id);

    /** Get all orders for a customer. */
    List<OrderResponse> getCustomerOrders(Long customerId);
}

File: frontend/src/services/customerService.ts
import api from './api';

export interface Customer {
    id: number;
    firstName: string;
    lastName: string;
    email: string;
    phone?: string;
    address?: string;
}

export interface CreateCustomerRequest {
    firstName: string;
    lastName: string;
    email: string;
    phone?: string;
    address?: string;
}

export interface UpdateCustomerRequest {
    firstName?: string;
    lastName?: string;
    email?: string;
    phone?: string;
    address?: string;
}

class CustomerService {
    private readonly endpoint = '/customers';

    async getAll(): Promise<Customer[]> {
        const response = await api.get<Customer[]>(this.endpoint);
        return response.data;
    }

    async getById(id: number): Promise<Customer> {
        const response = await api.get<Customer>(`${this.endpoint}/${id}`);
        return response.data;
    }

    async getByEmail(email: string): Promise<Customer> {
        const response = await api.get<Customer>(`${this.endpoint}/email/${email}`);
        return response.data;
    }

    async create(customerData: CreateCustomerRequest): Promise<Customer> {
        const response = await api.post<Customer>(this.endpoint, customerData);
        return response.data;
    }

    async update(id: number, customerData: UpdateCustomerRequest): Promise<Customer> {
        const response = await api.put<Customer>(`${this.endpoint}/${id}`, customerData);
        return response.data;
    }

    async delete(id: number): Promise<void> {
        await api.delete(`${this.endpoint}/${id}`);
    }
}

export default new CustomerService();
